### ISSUE TITLE
Make method effectivelyNotNull protected

### ISSUE THREAD

**Issue by kowy** [jOOQ/jOOQ#15963]
### Use case

We use JavaGenerator to generate java POJOs for our Kotlin application. We take advantage of how Kotlin work with platform types, so they are not nullable (and we don't have to place !! or ?. everywhere), but in special cases (POJO creation) we can leave null there. But this behaviour stopped working with changes in #10212. I understand you view and see that it is not possible to put flag to return old behaviour. But is it possible to change effectivelyNotNull method to protected so we can overwrite it in subclass?

### Possible solution

Make effectivelyNotNull method in JavaGenerator class protected so anybody can override it in own subclass and change how @Nullable/@NotNull annotations are generated.

### Possible workarounds

There is no other possible workaround

### jOOQ Version

jOOQ 3.18.7

### Database product and version

PostgreSQL 12.2 (Debian 12.2-2.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit

### Java Version

openjdk 17

### OS Version

Debian Linux

**Comment by lukaseder** [comment]
It's almost always not a good idea to make arbitrary internal methods overridable, because we'll break any "contract" that you might have assumed even in patch releases.

Specifically, you haven't illustrated how *exactly* your assumptions broke, so it doesn't seem wise to just tamper with internals to make sure your assumptions continue working.

Can you please state what exactly changed for you, by showing an example?

**Comment by kowy** [comment]
@lukaseder, thank you for your quick reply.

Let create simple example, how we use jOOQ's JavaGenerator. 

Example Postgres table:
```
create table survey (
  id integer primary key not null,
  description character varying,
  status survey_status_enum not null default 'UNDER_CONSTRUCTION'::survey_status_enum,
  language_code character varying not null default 'cs_CZ',
);
```
Until changes in #10212 took place, JavaGenerator generated this POJOs (but I think @Nullable/@NotNull annotations were generated same way also in Records, DAOs and so on):
```
public class Survey implements Serializable {
    ....
    public Survey(
        @NotNull Integer id,
        @Nullable String description,
        @NotNull SurveyStatusEnum status,
        @NotNull String languageCode,
    ) {
        this.id = id;
        this.description = description;
        this.status = status;
        this.languageCode = languageCode;
    }
    ....
}
```

So we can in Kotlin code use these lines:
```
val runningSurveyIds = allSurveys.filter { it.status.isRunning() }.map { it.id }.distinct()
val restPaylod = SurveyNewDto(
  surveyId = survey.id,
  type = survey.type.name,
  languageCode = survey.languageCode
)
// and still we can create new entities and not specify ID and defaulted columns
val survey = Survey().apply { description = "New survey" }
```

But after #10212 we have to put !! after each usage of Survey attributes (even though we know there can't be NULL value, because this object was read from DB):
```
val runningSurveyIds = allSurveys.filter { it.status!!.isRunning() }.map { it.id!! }.distinct()
val restPaylod = SurveyNewDto(
  surveyId = survey.id!!,
  type = survey.type.name!!,
  languageCode = survey.languageCode!!
)
```

Since now we have to add !! garbage on hundreds of places in our code to solve "semantic lie" in several places (new entities to be created have very limited area where they are used and for outer join we use custom POJOs, so we rather well know when there may be null values in these columns)

You may suggest us to not use @Nullable annotation at all, but having this on real NULLABLE columns is still valuable for us (there may be null in these columns if you directly load this entity from DB)

In general, I understand your position and reason why you changed behaviour of effectivelyNotNull method (null values during entity creation, outer joins). So I don't try to change your mind in this case, I just kindly ask to make effectivelyNotNull protected. Or somehow give us possibility how to change @Nullable/@NotNull generation strategy.

**Comment by lukaseder** [comment]
Thanks for the explanation.

I don't think that opening up the internals (and then hacking) will be better than just patching jOOQ, a point I've been making time and again. Tomorrow, we'll want to rename or refactor that method or whatever, and your assumptions will break again.

You can open up the method on your end, and recompile the class, and then this change only affects *you*. But asking for a protected method means that we'll have to maintain its contract for everyone, introducing tons of new assumptions and breakages in the future.

Alternatively, if you look at the code generation logic, it's not that complex to generate POJOs, so why not roll your own, entirely? Why rely on jOOQ for this particular aspect? I've written tons of code generators in the past for similar purposes. Just override `JavaGenerator::generatePojo` and you're all set.

However, the one thing that is different in your case compared to other similar cases is that you're using Java annotations in Java generated classes from kotlin, instead of kotlin generated classes. So, one option that (I think) has never been discussed in such contexts is to *omit* the annotation only on attributes where the write value can be `NULL` but the read value cannot (after a fetch). That would effectively result in the platform type `T!` again for these columns, while the others would continue to be `T` (always `NOT NULL`) or `T?` (always nullable). Would that suit your needs?

Note, the only really thorough answer to these problems has been discussed in this thread:

- https://github.com/jOOQ/jOOQ/issues/14566

In particular:

- https://github.com/jOOQ/jOOQ/issues/14566#issuecomment-1413851711

**Comment by kowy** [comment]
Thank you for your answer.

I don't think forking `JavaGenerator::generatePojo` makes any sense for us (or even for anybody). Because it depends on very complex methods like `generatePojo0`, `generatePojoSetters`, and so on. And these huge methods can't be forked because they are also private. And to be consistent, we also have to duplicate internals of `generateDao()`, `generateRecord()`, and ... I don't know what else. But we love your generator and found that 99,99% of the generated code (for DAOs, Records, and POJOs) is perfect for us. The only option you have given to us is to copy the whole JavaGenerator class, change two lines in `effectivelyNotNull` and build jooq-generator ourselves. And every time you make any change to JavaGenerator we have to copy these changes to the custom generator and rebuild it. What an ugly solution compared to forking and changing only a tiny `effectivelyNotNull` method. Therefore, the possibility that you change such a small contract in the future is still much better than following all your changes in the whole class.

Regarding your question about omitting annotations for `NOT NULL` attributes. Yes, this suggested option suits us in case you generate `@Nullable` annotation only in case that *read* value is really nullable (`type.isNullable() == true`). And omit the annotation on attributes where the *write* value can be `NULL`, but the read *value* cannot. To provide a complete picture, I expect `@NotNull` annotation would be generated on attributes when it is neither *Nullable* nor *Defaulted* nor *Identity*. This implementation would be perfect for our needs.

**Comment by lukaseder** [comment]
> Therefore, the possibility that you change such a small contract in the future is still much better than following all your changes in the whole class.

No dice. Internals won't be opened like that. People have asked similar questions many many times, and if I had said yes every time, there would only be regrets, even if you don't see that right now.

Anyway, I think the option of omitting any annotation on types where a write value can be nullable will be a much more thorough solution for this particular case, at least for users of `JavaGenerator`

**Comment by lukaseder** [comment]
I've created a new issue for this:

- https://github.com/jOOQ/jOOQ/issues/16002

I'm rejecting this issue here as I won't open up random internals.
